File: 백준/Silver/1992. 쿼드트리/쿼드트리.cc
#include <bits/stdc++.h>
using namespace std;
int n, a[70][70];
string ret;

string go(int y, int x, int size)
{
    if (size == 1)
    {
        return to_string(a[y][x]);
    }

    int now = a[y][x];
    bool flag = true;

    for (int i = y; i < y + size; i++)
    {
        for (int j = x; j < x + size; j++)
        {
            if (a[i][j] != now)
            {
                flag = false;
                break;
            }
        }
        if (!flag)
            break;
    }

    if (flag)
        return to_string(a[y][x]);

    int newSize = size / 2;
    string ret = "";
    ret += "(";
    ret += go(y, x, newSize);
    ret += go(y, x + newSize, newSize);
    ret += go(y + newSize, x, newSize);
    ret += go(y + newSize, x + newSize, newSize);
    ret += ")";
    return ret;
}
int main()
{
    scanf("%d", &n);

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            scanf("%1d", &a[i][j]);
        }
    }

    cout << go(0, 0, n) << "\n";
    return 0;
}

File: 백준/Silver/1992. 쿼드트리/쿼드트리.java
import java.util.Scanner;

public class Main {
    static int n;
    static int[][] a;

    public static String go(int y, int x, int siz) {
        String ret = "";
        boolean flag = false;

        int now = a[y][x];
        for (int i = y; i < y + siz; i++) {
            for (int j = x; j < x + siz; j++) {
                if (now != a[i][j]) {
                    flag = true;
                    break;
                }
            }
            if (flag)
                break;
        }

        if (!flag)
            return Integer.toString(now);
        else {
            ret += "(";
            ret += go(y, x, siz / 2);
            ret += go(y, x + siz / 2, siz / 2);
            ret += go(y + siz / 2, x, siz / 2);
            ret += go(y + siz / 2, x + siz / 2, siz / 2);
            ret += ")";
        }
        return ret;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();
        a = new int[n][n];

        for (int i = 0; i < n; i++) {
            String str = sc.next();
            for (int j = 0; j < n; j++) {
                a[i][j] = str.charAt(j) - '0';
            }
        }

        System.out.println(go(0, 0, n));
        sc.close();
    }
}
